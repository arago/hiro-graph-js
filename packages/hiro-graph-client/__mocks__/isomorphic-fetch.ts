/* eslint-env jest */
// mock of isomorphic-fetch for tests.
// fetch is a single function so we mock it with a
// mock fn.

// we need to require the real thing, as it mutates `global`, giving us access
// to Response
require.requireActual('isomorphic-fetch');

// @ts-ignore
const Response = global.Response;
const mockFn = jest.fn();

export { mockFn };

export default function fetch(...args: any[]) {
  // wrap the outcome of the mock function in a "new Response";
  const result = mockFn(...args);

  if (!result) {
    return Promise.resolve(new Response(JSON.stringify({ args })));
  }

  const [status, body] = result;

  if (body instanceof Error) {
    return Promise.reject(body);
  } else if (typeof body === 'string') {
    return Promise.resolve(new Response(body, { status }));
  } else {
    return Promise.resolve(new Response(JSON.stringify(body), { status }));
  }
}
