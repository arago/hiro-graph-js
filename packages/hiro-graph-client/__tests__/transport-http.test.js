/* eslint-env jest */
import { mockFn } from 'isomorphic-fetch';

import Client from '../src/client';
import clientTestHelper from '../__mocks__/client-test-helper.js';

const fakeTokenPromise = Promise.resolve('<token>');
const onInvalidate = jest.fn();
const fakeToken = { get: () => fakeTokenPromise, onInvalidate };

// These test are run with a client, in order to assure that the correct
// client calls result in the correct HTTP requests.
describe('transport-http', () => {
    // uses the real HttpTransport, but with fake fetch
    const client = new Client(
        { endpoint: 'https://graphit/', token: fakeToken },
        { forceHTTP: true },
    );

    // uncomment the next line to get a lot of debug output...
    //client.introspect(event => console.log(event));

    const tests = clientTestHelper(client);

    beforeEach(() => mockFn.mockReset());

    tests.forEach(({ name, fn }) => {
        it(`should create the right options for: ${name}`, async () => {
            // for now every request should recieve the same mock response.
            mockFn.mockReturnValueOnce([200, {}]);
            await fn();
            expect(mockFn).toHaveBeenCalledTimes(1);
            // @IMPORTANT check the snapshots! They must be correct the first time or these tests are worthless!
            expect(mockFn.mock.calls[0]).toMatchSnapshot(); //i.e. correct options created
        });
    });

    it('should handle bad http responses appropriately', async () => {
        const bad = [
            [200, 'Totally invalid response'],
            [403, { error: { code: 403, message: 'forbidden' } }],
            [404, { error: { code: 404, message: 'not found' } }],
        ];

        for (let i = 0; i < bad.length; i++) {
            mockFn.mockReturnValueOnce(bad[i]);

            let error;

            try {
                await client.me();
            } catch (e) {
                error = { code: e.code, message: e.message };
            }

            expect(error).toBeDefined();
            expect(error).toMatchSnapshot();
        }
    });
});
