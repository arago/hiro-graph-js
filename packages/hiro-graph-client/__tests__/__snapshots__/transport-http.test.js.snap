// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transport-http should create the right options for: connect 1`] = `
Array [
  "https://graphit/connect/some-verb",
  Object {
    "body": "{\\"in\\":\\"the-in-id\\",\\"out\\":\\"the-out-id\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: create (waitForIndex) 1`] = `
Array [
  "https://graphit/new/some-type?waitForIndex=true",
  Object {
    "body": "{\\"some\\":\\"prop\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: create 1`] = `
Array [
  "https://graphit/new/some-type",
  Object {
    "body": "{\\"some\\":\\"prop\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: delete (waitForIndex) 1`] = `
Array [
  "https://graphit/some-id",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "DELETE",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: delete 1`] = `
Array [
  "https://graphit/some-id",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "DELETE",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: disconnect 1`] = `
Array [
  "https://graphit/the-out-id%24%24some-verb%24%24the-in-id",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "DELETE",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: get 1`] = `
Array [
  "https://graphit/some-id",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: gremlin 1`] = `
Array [
  "https://graphit/query/gremlin",
  Object {
    "body": "{\\"root\\":\\"root-id\\",\\"query\\":\\"pow()\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with from + to + offset + limit) 1`] = `
Array [
  "https://graphit/some-id/history?offset=100&limit=5&from=0&to=1550597976759",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with from + to) 1`] = `
Array [
  "https://graphit/some-id/history?from=0&to=1550597976759",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with from) 1`] = `
Array [
  "https://graphit/some-id/history?from=0",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with limit) 1`] = `
Array [
  "https://graphit/some-id/history?limit=500",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with offset + limit) 1`] = `
Array [
  "https://graphit/some-id/history?offset=100&limit=5",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with offset) 1`] = `
Array [
  "https://graphit/some-id/history?offset=10",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with to) 1`] = `
Array [
  "https://graphit/some-id/history?to=1550597976759",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history (with version) 1`] = `
Array [
  "https://graphit/some-id/history?limit=500&version=123",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: history 1`] = `
Array [
  "https://graphit/some-id/history",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: ids 1`] = `
Array [
  "https://graphit/query/ids",
  Object {
    "body": "{\\"query\\":\\"a,b,c\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: info 1`] = `
Array [
  "https://graphit/info",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer false",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (basic) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"*:*\\",\\"limit\\":50,\\"offset\\":0}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (count) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"*:*\\",\\"limit\\":50,\\"offset\\":0,\\"count\\":\\"true\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (fields) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"*:*\\",\\"limit\\":50,\\"offset\\":0,\\"fields\\":\\"/field1,/field2\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (limits) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"*:*\\",\\"limit\\":-1,\\"offset\\":1000}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (order) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"*:*\\",\\"limit\\":50,\\"offset\\":0,\\"order\\":\\"/field1 asc,/field2 desc\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: lucene (placeholders) 1`] = `
Array [
  "https://graphit/query/vertices",
  Object {
    "body": "{\\"query\\":\\"/foo:$bar\\",\\"limit\\":50,\\"offset\\":0,\\"bar\\":\\"123\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: me 1`] = `
Array [
  "https://graphit/_me",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: replace (createIfNotExists) 1`] = `
Array [
  "https://graphit/some-id?createIfNotExists=true&ogit%2F_type=some-type",
  Object {
    "body": "{\\"/foo\\":\\"bar\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "PUT",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: replace (waitForIndex) 1`] = `
Array [
  "https://graphit/some-id?waitForIndex=true",
  Object {
    "body": "{\\"/foo\\":\\"bar\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "PUT",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: replace 1`] = `
Array [
  "https://graphit/some-id",
  Object {
    "body": "{\\"/foo\\":\\"bar\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "PUT",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: streamts (from only) 1`] = `
Array [
  "https://graphit/ts-id/values?limit=50&from=1399999999999",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: streamts (to + from) 1`] = `
Array [
  "https://graphit/ts-id/values?limit=50&from=1399999999999&to=1400000000001",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: streamts (to only) 1`] = `
Array [
  "https://graphit/ts-id/values?limit=50&to=1400000000001",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: streamts 1`] = `
Array [
  "https://graphit/ts-id/values?limit=50",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: update (waitForIndex) 1`] = `
Array [
  "https://graphit/some-id?waitForIndex=true",
  Object {
    "body": "{\\"/foo\\":\\"bar\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: update 1`] = `
Array [
  "https://graphit/some-id",
  Object {
    "body": "{\\"/foo\\":\\"bar\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should create the right options for: writets 1`] = `
Array [
  "https://graphit/ts-id/values",
  Object {
    "body": "{\\"items\\":[{\\"timestamp\\":1400000000000,\\"value\\":\\"value1\\"}]}",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer <token>",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "mode": "cors",
  },
]
`;

exports[`transport-http should handle bad http responses appropriately 1`] = `
Object {
  "code": 502,
  "message": "[HTTP] Error: Invalid JSON in response from GraphIT",
}
`;

exports[`transport-http should handle bad http responses appropriately 2`] = `
Object {
  "code": 403,
  "message": "[HTTP] Error: forbidden",
}
`;

exports[`transport-http should handle bad http responses appropriately 3`] = `
Object {
  "code": 404,
  "message": "[HTTP] Error: not found",
}
`;
